src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -0,0 +1,346 @@
+    default:
+        ChipLogError(DataManagement, "Attribute type 0x%x not handled", static_cast<int>(metadata->attributeType));
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedWrite);
+    }
+    return CHIP_IM_GLOBAL_STATUS(UnsupportedWrite);
+}
+
+} // namespace
+
+CHIP_ERROR CodegenDataModel::WriteAttribute(const InteractionModel::WriteAttributeRequest & request,
+                                            AttributeValueDecoder & decoder)
+{
+    ChipLogDetail(DataManagement,
+                  "Reading attribute: Cluster=" ChipLogFormatMEI " Endpoint=%x AttributeId=" ChipLogFormatMEI " (expanded=%d)",
+                  ChipLogValueMEI(request.path.mClusterId), request.path.mEndpointId, ChipLogValueMEI(request.path.mAttributeId),
+                  request.path.mExpanded);
+
+    // ACL check for non-internal requests
+    if (!request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+        ReturnErrorCodeIf(!request.subjectDescriptor.has_value(), CHIP_ERROR_INVALID_ARGUMENT);
+
+        Access::RequestPath requestPath{ .cluster = request.path.mClusterId, .endpoint = request.path.mEndpointId };
+        ReturnErrorOnFailure(Access::GetAccessControl().Check(*request.subjectDescriptor, requestPath,
+                                                              RequiredPrivilege::ForWriteAttribute(request.path)));
+    }
+
+    auto metadata = Ember::FindAttributeMetadata(request.path);
+
+    // Explicit failure in finding a suitable metadata
+    if (const CHIP_ERROR * err = std::get_if<CHIP_ERROR>(&metadata))
+    {
+        VerifyOrDie(*err != CHIP_NO_ERROR);
+        return *err;
+    }
+
+    // All the global attributes that we do not have metadata for are
+    // read-only (i.e. cannot write atribute_list/event_list/accepted_cmds/generated_cmds)
+    //
+    // so if no metadata available, we wil lreturn an error
+    const EmberAfAttributeMetadata ** attributeMetadata = std::get_if<const EmberAfAttributeMetadata *>(&metadata);
+    if (attributeMetadata == nullptr)
+    {
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedWrite);
+    }
+
+    if ((*attributeMetadata)->IsReadOnly() && !request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+        // Internal is allowed to try to bypass read-only updates, however otherwise we deny read-only
+        // updates
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedWrite);
+    }
+
+    if ((*attributeMetadata)->MustUseTimedWrite() && !request.writeFlags.Has(InteractionModel::WriteFlags::kTimed))
+    {
+        return CHIP_IM_GLOBAL_STATUS(NeedsTimedInteraction);
+    }
+
+    if (request.path.mDataVersion.HasValue())
+    {
+        std::optional<InteractionModel::ClusterInfo> clusterInfo = GetClusterInfo(request.path);
+        if (!clusterInfo.has_value())
+        {
+            ChipLogError(DataManagement, "Unable to get cluster info for Endpoint %x, Cluster " ChipLogFormatMEI,
```

Comment:
How can this happen? By now, a critical invariant is that we need to have all cluster data available. I think it's safer not to crash here, but this is not a DataVersionMismatch, this is an internal error.

Reply:
Updated to CHIP_ERROR_NOT_FOUND (even though realistically this is terrible ... but also likely not possible in the existing code). I carried it over from:

https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/ember-compatibility-functions.cpp#L168 which returns CHIP_ERROR_NOT_FOUND